tooling/nargo_cli/src/cli/test_cmd.rs

```diff
@@ -68,258 +94,410 @@ pub(crate) fn run(args: TestCommand, config: NargoConfig) -> Result<(), CliError
         None => FunctionNameMatch::Anything,
     };
 
-    // Configure a thread pool with a larger stack size to prevent overflowing stack in large programs.
-    // Default is 2MB.
-    let pool = rayon::ThreadPoolBuilder::new().stack_size(4 * 1024 * 1024).build().unwrap();
-    let test_reports: Vec<Vec<(String, TestStatus)>> = pool.install(|| {
-        workspace
-            .into_iter()
-            .par_bridge()
-            .map(|package| {
-                run_tests::<Bn254BlackBoxSolver>(
-                    &workspace_file_manager,
-                    &parsed_files,
-                    package,
-                    pattern,
-                    args.show_output,
-                    args.oracle_resolver.as_deref(),
-                    Some(workspace.root_dir.clone()),
-                    Some(package.name.to_string()),
-                    &args.compile_options,
-                )
-            })
-            .collect::<Result<_, _>>()
-    })?;
-    let test_report: Vec<(String, TestStatus)> = test_reports.into_iter().flatten().collect();
-
-    if test_report.is_empty() {
-        match &pattern {
-            FunctionNameMatch::Exact(pattern) => {
-                return Err(CliError::Generic(
-                    format!("Found 0 tests matching input '{pattern}'.",),
-                ))
+    let num_threads = args.test_threads.unwrap_or_else(rayon::current_num_threads);
+
+    let runner = TestRunner {
+        file_manager: &file_manager,
+        parsed_files: &parsed_files,
+        workspace,
+        args: &args,
+        pattern,
+        num_threads,
+    };
+    runner.run()
+}
+
+struct TestRunner<'a> {
+    file_manager: &'a FileManager,
+    parsed_files: &'a ParsedFiles,
+    workspace: Workspace,
+    args: &'a TestCommand,
+    pattern: FunctionNameMatch<'a>,
+    num_threads: usize,
+}
+
+impl<'a> TestRunner<'a> {
+    fn run(&self) -> Result<(), CliError> {
+        // First compile all packages and collect their tests
+        let package_test_results = self.collect_packages_tests();
+
+        // Now gather all tests and how many are per packages
+        let mut tests = Vec::new();
+        let mut test_count_per_package = HashMap::new();
+
+        for (package_name, package_tests) in package_test_results {
+            let package_tests = package_tests?;
```

Comment:
So we get the report from the first package that failed to compile, but not the rest? Or did it already print something about why they failed earlier?

Reply:
Right, the way it works is a bit confusing: the error is printed right away and `Err` is returned. So by the time we get here all the errors were already printed 🤷

Reply:
That is, it's confusing but it was already working like that (it's how compilation works)

Reply:
Then maybe return `Result<BTreeMap<String, Vec<Test<'a>>>, CliError>` from `collect_packages_tests` to indicate that there's nothing more to do with this?

Reply:
Done! Though I'm sure there's a more idiomatic Rust way to do this...

Reply:
With stuff like `Vec<Result<_, _>>` you can use `let foo: Result<Vec<_>, _> = results.collect();` but I'm not sure it works with a `BTreeMap`.